rfc/20250317-ephemeral-resources.md

```diff
@@ -0,0 +1,646 @@
+  ```hcl
+    │ Error: An output referencing a sensitive value needs to be marked with sensitive too
+    │
+    │   on main.tf line 32:
+    │   32: output "write_only_out" {
+    │
+    │ For security reasons, OpenTofu requires any output that is referencing a sensitive value to also be configured the same. If the root module really wants to export this sensitive value, you need to annotate it with the following argument:
+    │     sensitive = true
+  ```
+
+Considering the rules above, root modules cannot have any ephemeral outputs defined.
+
+### Locals
+Any `local` declaration will be marked as ephemeral if in the expression that initialises it an ephemeral value is used:
+```hcl
+variable "var1" {
+  type = string
+}
+
+variable "var2" {
+  type = string
+}
+
+variable "var3" {
+  type      = string
+  ephemeral = true
+}
+
+locals {
+  eg1 = var.var1 == "" ? var.var2 : var.var1 // not ephemeral
+  eg2 = var.var2 // not ephemeral
+  eg3 = var.var3 == "" ? var.var2 : var.var1 // ephemeral because of var3 conditional
+  eg4 = var.var1 == "" ? var.var2 : var.var3 // ephemeral because of var3 usage
+  eg5 = "${var.var3}-${var.var1}" // ephemeral because of var3 usage
+  eg6 = local.eg4 // ephemeral because of eg4 is ephemeral
+}
+```
+
+Once a local is marked as ephemeral, this can be used only in other ephemeral contexts. Check the `User Documentation` section for more details on the allowed contexts.
+
+### Ephemeral resources
+Due to the fact ephemeral resources are not stored in the state/plan file, this block is not creating a diff in the OpenTofu's UI.
+Instead, OpenTofu will notify the user of opening/renewing/closing an ephemeral resource with messages similar to the following:
+```bash
+ephemeral.playground_random.password: Opening...
+ephemeral.playground_random.password: Opening succeeded after 0s
+ephemeral.playground_random.password: Closing...
+ephemeral.playground_random.password: Closing succeeded after 0s
+```
+
+Ephemeral resources lifecycle is similar with the data blocks:
+* Both basic implementations require the same methods (`Metadata` and `Schema`) while datasource is defining `Read` compared with the ephemeral resource that is defining `Open`. When talking about the basic functionality of the ephemeral resources, the `Read` method will behave similarly with the `Read` on a datasource, where it reads the data.
+* Also, both blocks support `Configure`, `ConfigValidators` and `ValidateConfig` as extensions of the basic definition.
+* Ephemeral resources do support two more operations in contrast with datasources:
+  * `Renew`
+    * Together with the data returned by the `Open` method call, the provider can also specify a `RenewAt` which will be a specific moment in time when OpenTofu should call the `Renew` method to get an updated information from the ephemeral resource. OpenTofu will have to check for `RenewAt` value anytime it intends to use the value returned by the ephemeral resource.
```

Comment:
This renewat functionality isn't super clear to me based on the doc above, I wasn't expecting this. 

Can you describe it in abit more detail and why it's needed? Is it so that providers can determine if/when opentofu should refetch rather than it only being a human (version) driven re-fetch?

Reply:
Is the `Renew` [method] (https://github.com/opentofu/opentofu/blob/6b81cb76c65cb76311db9ecf9d67a94fbb0185cc/rfc/20250317-ephemeral-resources.md#renew-method-details) dedicated section enough?